### PR TITLE
Make Deployer Testable

### PR DESCRIPTION
**Pull request by breck7**
Still need to clean it up a bit more and then add the tests, but this should be a big step toward a lot better deploys/bakes/etc.

- [x] Step 1: bash to Typescript
- [x] Step 2: split one long script to units
- [x] Step 3: progress bars and benchmarking of units
- [x] Step 4: break code up into subprojects
- [x] Step 5: make deployer testable
- [ ] Step 6: more cleanup and add tests
- [ ] Step 7: super fast admin deploys

### CODE REVIEW

File: baker/Deployer.ts [hunk 1]
```diff
@@ -0,0 +1,372 @@
+#! /usr/bin/env node
```
Comment (breck7) [inline baker/Deployer.ts:1]: nit: remove. this one isnt executable anymore.

File: baker/Deployer.ts [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+import * as prompts from "prompts"
+import ProgressBar = require("progress")
+import { execWrapper } from "../db/execWrapper"
+import { spawn } from "child_process"
+import simpleGit, { SimpleGit } from "simple-git"
+import { WriteStream } from "tty"
+import { ProgressStream } from "./ProgressStream"
+
+const TEMP_DEPLOY_SCRIPT_SUFFIX = `.tempDeployScript.sh`
+
```
Comment (breck7) [inline baker/Deployer.ts:13]: these things are fugly. clean up

File: baker/Deployer.ts [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+            console.log(`Exit code: ${exitCode}`)
+        }
+
+        this.progressBar.tick({
+            name: `✅ 📡 finished running ${path}`,
+        })
+    }
+}
+
+const makeScriptToDoFileStuff = (
```
Comment (breck7) [inline baker/Deployer.ts:312]: these methods are clearly perfect. no future work needed.

File: baker/ProgressStream.ts [hunk 4]
```diff
@@ -0,0 +1,36 @@
+import { WriteStream } from "tty"
+
+// Wrap stderr before passing it to ProgressBar so we can save all writes
+// and replay them at the end of the bake. Without this the progress bar class
+// works fine, but there is no way to show the summary once the job is complete.
+export class ProgressStream implements Partial<WriteStream> {
```
Comment (breck7) [inline baker/ProgressStream.ts:6]: should probably share this upstream with progress bar?

File: baker/Deployer.ts [hunk 5]
```diff
@@ -0,0 +1,374 @@
...
+import ProgressBar = require("progress")
+import { execWrapper } from "../db/execWrapper"
+import { spawn } from "child_process"
+import simpleGit, { SimpleGit } from "simple-git"
+import { WriteStream } from "tty"
+import { ProgressStream } from "./ProgressStream"
+
+const TEMP_DEPLOY_SCRIPT_SUFFIX = `.tempDeployScript.sh`
+
+enum DeployTarget {
```
Comment (MarcelGerber) [inline baker/Deployer.ts:12]: should probably name this something else (`StagingDeployTarget`?) to indicate that it doesn't include `live`.

Also we could maybe benefit by having that enum in its own file, or at least exporting it? https://github.com/owid/owid-grapher/blob/master/deploy/liveCommit.ts also uses it.

File: baker/Deployer.ts [hunk 6]
```diff
@@ -0,0 +1,374 @@
...
+        this.progressBar.tick({ name: `✅  finished ${command}` })
+    }
+
+    private get targetIsStaging() {
+        return new Set(Object.values(DeployTarget)).has(this.options.target)
+    }
+
+    get targetIsProd() {
+        return this.options.target.toString() === "live"
+    }
```
Comment (MarcelGerber) [inline baker/Deployer.ts:67]: The type of `this.options.target` is `DeployTarget`, so I guess that shouldn't even allow for `live`?

File: baker/Deployer.ts [hunk 7]
```diff
@@ -0,0 +1,374 @@
...
+interface DeployerOptions {
+    owidGrapherRootDir: string
+    userRunningTheDeploy: string
+    target: DeployTarget
+    skipChecks?: boolean
+    runChecksRemotely?: boolean
+}
+
+const OWID_STAGING_DROPLET_IP = "165.22.127.239"
+const OWID_LIVE_DROPLET_IP = "209.97.185.49"
```
Comment (MarcelGerber) [inline baker/Deployer.ts:33]:
```suggestion
const OWID_STAGING_SERVER_IP = "165.22.127.239"
const OWID_LIVE_SERVER_IP = "209.97.185.49"
```

... so as to not get too into it with the DigitalOcean terminology

File: baker/Deployer.ts [hunk 8]
```diff
@@ -0,0 +1,374 @@
...
+        Object.keys(scripts).forEach((name) => {
+            const fullName = `${name}${TEMP_DEPLOY_SCRIPT_SUFFIX}`
+            const localPath = `${owidGrapherRootDir}/${fullName}`
+            fs.writeFileSync(localPath, scripts[name], "utf8")
+            fs.chmodSync(localPath, "755")
+        })
+
+        await this.copyLocalRepoToServerTmpDirectory()
+
+        for await (const name of Object.keys(scripts)) {
```
Comment (MarcelGerber) [inline baker/Deployer.ts:249]: nit: I don't think you need that `await`

File: baker/Deployer.ts [hunk 9]
```diff
@@ -0,0 +1,374 @@
...
+            await this.runAndStreamScriptOnRemoteServerViaSSH(
+                `${SYNC_TARGET}/${name}${TEMP_DEPLOY_SCRIPT_SUFFIX}`
+            )
+        }
+    }
+
+    printAndExit(message: string) {
+        // eslint-disable-next-line no-console
```
Comment (MarcelGerber) [inline baker/Deployer.ts:257]: Yeah I think we should get rid of that eslint rule, at least for server-side / node code...

File: baker/Deployer.ts [hunk 10]
```diff
@@ -0,0 +1,374 @@
...
+        const child = spawn(`ssh`, params)
+
+        child.stdout.on("data", (data) => {
+            // eslint-disable-next-line no-console
+            console.log(data.toString())
+        })
+
+        child.stderr.on("data", (data) => {
+            // eslint-disable-next-line no-console
+            console.log(data.toString())
```
Comment (MarcelGerber) [inline baker/Deployer.ts:296]: probably better to use `console.error` here to differentiate that output?

**Review by MarcelGerber — APPROVED** [review]
Looking good! But I had a few comments when going through it, so here they are.

File: baker/Deployer.ts (hunk 3 above)
Comment (breck7) [reply baker/Deployer.ts:314]: Cleaned them up a little but still work to be done there.

File: baker/Deployer.ts (hunk 5 above)
Comment (breck7) [reply baker/Deployer.ts]: Good call. Done and done.

File: baker/Deployer.ts (hunk 6 above)
Comment (breck7) [reply baker/Deployer.ts:56]: You're right. That was ugly. Just cleaned that up.

File: baker/Deployer.ts (hunk 7 above)
Comment (breck7) [reply baker/Deployer.ts:22]: I am a digital ocean fanboy. While I agree with you, and wouldn't object if someone else makes this change, I cannot personally bring myself to do it.

File: baker/Deployer.ts (hunk 9 above)
Comment (breck7) [reply baker/Deployer.ts:246]: Hmm, I kind of like it? It has stopped me from leaving in console junk a number of times.

File: baker/Deployer.ts (hunk 10 above)
Comment (breck7) [reply baker/Deployer.ts:285]: Hmm, probably? I know some stuff is printing to stderr that maybe shouldn't. (I think progress bar? forget which). I am not crazy about `stderr` and `stdout` (just like `null` and `undefined`, really what you want is `N` standard channels).

**Pull request merged**